File: equistore/tests/serialization.rs
#[test]
fn load_file() {
    let tensor = equistore::io::load("./tests/data.npz").unwrap();

    assert_eq!(tensor.keys().names(), ["spherical_harmonics_l", "center_species", "neighbor_species"]);
    assert_eq!(tensor.keys().count(), 27);

    let block = tensor.block_by_id(13);

    assert_eq!(block.values().as_array().shape(), [9, 3, 3]);
    assert_eq!(block.samples().names(), ["structure", "center"]);
    assert_eq!(block.components().len(), 1);
    assert_eq!(block.components()[0].names(), ["spherical_harmonics_m"]);
    assert_eq!(block.properties().names(), ["n"]);

    assert_eq!(block.gradient_list(), ["positions"]);
    let gradient = block.gradient("positions").unwrap();
    assert_eq!(gradient.values().as_array().shape(), [27, 3, 3, 3]);
    assert_eq!(gradient.samples().names(), ["sample", "structure", "atom"]);
    assert_eq!(gradient.components().len(), 2);
    assert_eq!(gradient.components()[0].names(), ["gradient_direction"]);
    assert_eq!(gradient.components()[1].names(), ["spherical_harmonics_m"]);
    assert_eq!(gradient.properties().names(), ["n"]);
}

File: python/equistore-torch/tests/serialization.py
import os

import pytest
import torch

import equistore.torch

from . import utils


@pytest.fixture
def tensor():
    return utils.tensor(dtype=torch.float64)


def test_load():
    tensor = equistore.torch.load(
        os.path.join(
            os.path.dirname(__file__),
            "..",
            "..",
            "..",
            "equistore",
            "tests",
            "data.npz",
        ),
    )

    assert tensor.keys.names == [
        "spherical_harmonics_l",
        "center_species",
        "neighbor_species",
    ]
    assert len(tensor.keys) == 27

    block = tensor.block(
        dict(spherical_harmonics_l=2, center_species=6, neighbor_species=1)
    )
    assert block.samples.names == ["structure", "center"]
    assert block.values.shape == (9, 5, 3)

    gradient = block.gradient("positions")
    assert gradient.samples.names == ["sample", "structure", "atom"]
    assert gradient.values.shape == (59, 3, 5, 3)


def test_save(tmpdir, tensor):
    """Check that we can save and load a tensor to a file"""
    tmpfile = "serialize-test.npz"

    with tmpdir.as_cwd():
        equistore.torch.save(tmpfile, tensor)
        data = equistore.torch.load(tmpfile)

    assert len(data.keys) == 4
